### ISSUE TITLE
How to setup debugging for a love2d application

### ISSUE THREAD

**Issue by bploeckelman** [EmmyLua/Intellij-EmmyLua2#6]
The original EmmyLua IntelliJ plugin has limitations and a couple bugs that make it a bit frustrating to use, such as syntax highlighting failing intermittently, environment variables not being correctly persisted in run configurations without a manual workaround, and muted breakpoints sometimes still suspending execution. EmmyLua2 is an improvement, but I can't figure out how to debug a love2d application.

I assume "Launch Debugger" should be used to start the program to be debugged and "Attach Debugger" connects to the running application process to enable breakpoints and other debugging features. But this isn't working for me on Windows 11 with a Love2d application that I can debug using EmmyLua version 1.

Launch Debugger Configuration:
- Program: `C:\path\to\love.exe`
- Working Directory: `C:\path\to\project\root`
- Parameters: `.`

This is equivalent to what I've used in the "Lua Application" run configuration from EmmyLua version 1. Love2d expects to be launched like this: `cd C:\path\to\project\root; C:\path\to\love.exe .`

Attach Debugger Configuration:
- Process Id: `0`
- Process Name: `love`
- Encoding: `gbk` (default, not sure what this controls or what the options are)
- (x) use name

Running "Launch Debugger", the debug window's console shows: "[F]Successfully inject dll! Try connect localhost:6193 ..." but the application doesn't start, or at least the window doesn't appear, there is a love.exe process running.

Then running "Attach Debugger" I see the following:
```shell
[F]The process already attached.
Try connect ::1:9269 ...
Connected.
analyze: love.exe
analyze: lua51.dll
	[B]Lua symbol: luaJIT_profile_dumpstack
	... more lua symbols...
	[B]Lua symbol: luaJIT_version_2_1_1700008891
	... more lua symbols...
analyze: MSVCP120.dll
analyze: love.dll
	[B]Lua symbol: luaopen_love
	[B]Lua symbol: luaopen_love_arg
	[B]Lua symbol: luaopen_love_audio
	... more love2d symbols...
analyze: MSVCR120.dll
analyze: SDL2.dll
analyze: OpenAL32.dll
analyze: mpg123.dll
```

So it seems like things are launching and connected as expected, but the application isn't running, or at least isn't visible.

Looking into the processes with [process explorer](https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer) I see a love.exe process running with the expected working directory and command and an emmy_tool.exe process with what seems like a mostly correct configuration, but there might be some issues with quoting in the command invocation.

love.exe process:
```shell
# Path:
C:\path\to\love.exe
# Command line:
"C:\path\to\love.exe" "".""
# Current directory: 
C:\path\to\project\root\
```
Clicking "Bring to Front" for the love.exe process shows a dialog with the message "No visible windows found for this process", which seems relevant.

emmy_tool.exe process:
```shell
# Path:
C:\Users\Username\AppData\Roaming\JetBrains\IntelliJIdea2024.2\plugins\EmmyLua2-AttachDebugger\debugger\bin\win32-x64\emmy_tool.exe
# Command line:
C:\Users\Username\AppData\Roaming\JetBrains\IntelliJIdea2024.2\plugins\EmmyLua2-AttachDebugger\debugger\bin\win32-x64\emmy_tool.exe launch -dll emmy_hook.dll -dir "\"C:/Users/Username/AppData/Roaming/JetBrains/IntelliJIdea2024.2/plugins/EmmyLua2-AttachDebugger/debugger/bin/win32-x64\"" -work "\"C:\path\to\project\root\"" -exe "\"C:\path\to\love.exe\"" -args "\".\""
# Current directory: 
C:\Users\Username\AppData\Roaming\JetBrains\IntelliJIdea2024.2\plugins\EmmyLua2-AttachDebugger\debugger\bin\win32-x64\
```

**Comment by CppCXY** [comment]
Essentially, EmmyLua2 is a port of vscode-emmylua. Try vscode-emmylua to see if you can debug correctly.

**Comment by CppCXY** [comment]
Because EmmyLua2 has removed most of the original EmmyLua code, some features have not been ported or have been ported incorrectly. I don't have experience with love2d development, but if you can guide me on how to run love2d, I might be able to give it a try.

**Comment by bploeckelman** [comment]
Thanks for the response! I'll see if I can get things running in vscode, but its not my primary IDE so I'll need some time to get familiar with it. In the meantime I've set up a demo project that should hopefully be straightforward to setup and test with: 

- https://github.com/bploeckelman/emmylua-love2d

The instructions for setting up the run configs in intellij are in the project's `src/main.lua` file. The only prerequisite should be downloading the [Love2d executable](https://love2d.org/) for your platform.

Happy to help further, just let me know if you have questions. Thanks again for your work on these plugins!

**Comment by CppCXY** [comment]
it work well with vscode:
![image](https://github.com/user-attachments/assets/6d79f274-1f10-4f24-98ca-7e803d2ca2d5)
![image](https://github.com/user-attachments/assets/2c96caeb-5690-4b41-a4d5-bf7bd9846898)

config:

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "type": "emmylua_launch",
            "request": "launch",
            "name": "Attach by launch program",
            "program": "${workspaceFolder}/love-11.5-win64/love.exe",
            "workingDir": "${workspaceFolder}",
            "arguments": [
                "${workspaceFolder}/src"
            ],
            "newWindow": true
        }
    ]
}
```

**Comment by CppCXY** [comment]
There might be an issue with attach debugger on IntelliJ or passing command line arguments. I will check and try to fix it when I have time.

**Comment by CppCXY** [comment]
Currently, I am busy implementing the language server using native languages. see: https://github.com/CppCXY/emmylua-analyzer-rust

**Comment by CppCXY** [comment]
I have fixed the debug issue on IntelliJ. You can wait for the plugin review and update.
![image](https://github.com/user-attachments/assets/b3cbea2d-cca1-4d3c-a33b-26834b6e76d1)

config:
programe: "path/to/love.exe"
working Directory: "path/to/source"
parameter: src

**Comment by bploeckelman** [comment]
I pulled the updated plugin and can confirm that these changes resolve the issue for me. I'm able to launch a love2d application via the `Emmy Launch Debugger` run configuration, attach the debugger with the `Emmy Attach Debugger` run configuration, hit breakpoints, step through, and examine variables.

For future reference, here's how my run configs are setup:

`Emmy Launch Debugger` config:
- Program: `path/to/love.exe`
- Working Directory: `path/to/love2d/project/source`
- Parameters: `.`

`Emmy Attach Debugger` config:
- Process Id: `0`
- Process Name: `love`
- Encoding: `gbk`
- ✔️ `use name`

Closing this issue as resolved.

Thanks again for your help, really appreciate your work on these plugins!
